### ISSUE TITLE
Tutorial 12: Vertical and horizontal convolution stacks

### ISSUE THREAD

**Issue by WNZhao1988** [phlippe/uvadlc_notebooks#47]
Hi,
Thanks for sharing such a great notebook! I have a question about the vertical and horizontal convolution stacks in tutorial 12. Based on your explanation:

> The vertical convolution is not allowed to work on features from the horizontal convolution. In the feature map of the horizontal convolutions, a pixel contains information about all of the "true" pixels on the left. If we apply a vertical convolution which also uses features from the right, we effectively expand our receptive field to the true input which we want to prevent. Thus, the feature maps can only be merged for the horizontal convolution.

I'm still confused about why for horizontal convolution we need to add horiz_conv(horiz_img) + vert_img but for vertical convolution, we only need vert_conv(vert_img).

Would appreciate if you can explain more about this!

**Comment by phlippe** [comment]
Hi @WNZhao1988, thanks for your kind words and your question! The easiest way of understanding this difference is to simulate your suggestion of adding `horiz_img` also to `vert_img`. The resulting receptive field for a center pixel is shown below:
![Screenshot from 2022-08-09 11-01-55](https://user-images.githubusercontent.com/25037725/183609442-c783b32a-435a-43df-b101-cf1249e19964.png)
As you might directly notice, the center pixel itself is part of the receptive field. In other words, to predict a pixel at position `(i,j)` (row `i`, column `j`), we look at the actual input at position `(i,j)`, which, during training, is the true pixel. Hence, we cannot really train this way.

So why does this happen? The reason is that the vertical convolution also contains weights for the pixel right of the center pixel (excluding the ones with `mask_center=True`). In other words, convolutions for pixel `(i,j)` also consider features at `(i,j+1)`. However, the features of a pixel from the horizontal convolutions contain all information from pixels left of it. This means that features at pixel `(i,j+1)` contain information of the original pixel at `(i,j)`. Since the vertical convolutions include pixel `(i,j+1)` in their direct receptive field, adding the horizontal features to the vertical block would break the autoregressive assumption of the model. In the extreme, i.e., after several vertical+horizontal convolutions, you would end up with considering all pixels in the same or earlier rows for predicting a pixel, which includes the pixel itself and all other pixels in the row that we would consider as "future" pixels that the model should not have access to at this stage. 

I hope this clarifies the difference, let me know if you have any questions remaining!

**Comment by WNZhao1988** [comment]
>

**Comment by WNZhao1988** [comment]
Hi @phlippe, thank you for your quick and detailed explanation! I think now I totally understand. 
In fact, I realized that previously I missed that in your code the mask for vertical convolution is  `mask[kernel_size//2+1:,:] = 0 ` instead of what I though mask[kernel_size//2:,:] = 0. 

```
class VerticalStackConvolution(MaskedConvolution):

    def __init__(self, c_in, c_out, kernel_size=3, mask_center=False, **kwargs):
        # Mask out all pixels below. For efficiency, we could also reduce the kernel
        # size in height, but for simplicity, we stick with masking here.
        mask = torch.ones(kernel_size, kernel_size)
        **mask[kernel_size//2+1:,:] = 0**

        # For the very first convolution, we will also mask the center row
        if mask_center:
            mask[kernel_size//2,:] = 0

        super().__init__(c_in, c_out, mask, **kwargs)
```

**Comment by phlippe** [comment]
Hi @WNZhao1988, indeed, we only need to mask the center line for the very first convolution to not look at the actual input pixel. Afterwards, we want to reuse our features for this pixel, hence only masking `mask[kernel_size//2+1:,:] = 0`.
I'm closing the issue, but feel free to reopen it in case questions remain.
